sklearn/preprocessing/_encoders.py

```diff
@@ -697,7 +709,23 @@ def transform(self, X):
         X_out : sparse matrix or a 2-d array
             Transformed input.
         """
-        X_int, _ = self._transform(X)
+        X_int, X_mask = self._transform(X, handle_unknown=self.unknown_value)
+
+        # create separate category for unknown values
+        if self.unknown_value != 'error':
+            if not isinstance(self.unknown_value, np.int):
+                raise TypeError("The used value for unknown_value {} is not "
+                                "an integer as required."
+                                .format(self.unknown_value))
+            n_features = len(self.categories_)
+            for i in range(n_features):
+                if np.isin(self.unknown_value, np.unique(X_int[:, i])):
```

Comment:
shouldn't we just be checking if it's in the range 0 to `len(self.categories_[i])`? I.e. if we're going to say this should raise an error, it should raise it even if this *specific* dataset does not contain an instance of that category.

Reply:
you are right. nice catch, thanks. fixed.